### ISSUE TITLE
direct drivable area segmentation with yolact

### ISSUE THREAD

**Issue by pandamax** [dbolya/yolact#92]
Thanks for your wonderful job！I have implemented direct drivable area segmentation with yolact_darknet53_config.And the test video below shows that it really works.But for the speed,is there any solution that use more lightweight model to be the backbone,for example [MobileNets?](https://arxiv.org/abs/1905.02244) 
[https://youtu.be/ZzLeWhUnpmI](url)

**Comment by dbolya** [comment]
Some people in #27 are trying out mobilenet, but I think you already saw that. Other than that, no I don't have plans to implement backbones other than the 4 we already have. However, adding your own backbone should be pretty easy. Check out the code in `backbone.py` and you should be able to follow along. If you do decide to try to implement it yourself, feel free to ask questions if you're stuck.

Though, mobilenet (because it's so lightweight) might actually struggle a lot with YOLACT, because we leave localization to the network.

Actually--if your application is drivable area segmentation, wouldn't it be better to use a semantic segmentation method? YOLACT makes compromises to be able to do instance segmentation, so if you don't need instance segmentation you're much better off just using semantic segmentation. (Though correct me if I'm wrong. I'm not familiar with drivable area segmentation)

**Comment by pandamax** [comment]
Yep!Thanks for your prompt reply.As you said, it's better to use semantic segmentation methed  to implement coarse drivable area segmentation especially taking speed into account.

But other classes will be added in the forthcoming experiment,for example alternative drivable area and car.Semantic segmetation method is hard to find the every lane‘s boundary when number of lane and driving car’s perspective is uncertain.So I think Instance segmentation method should be adopted.

Above all,thanks again for your contribution.

**Comment by dbolya** [comment]
That makes sense. By the way, if you want it, there's actually a semantic segmentation branch in YOLACT used only for training. See #73 for where you can expose that branch.

**Comment by pandamax** [comment]
Thank you for the quick response!
I will try the experiment based on your instance segmentation method when alternative drivable area is added.Then test if the model works.of course,I will try the method that you suggest.
Thanks again~

**Comment by pandamax** [comment]
Here is the test image, green area is the direct drivable area;and the other is the alternative drivable area.it seems fine,of course it depends the labeled datasets 
![2019-07-10 18-23-58屏幕截图](https://user-images.githubusercontent.com/23495527/60962022-f0d0e580-a33f-11e9-813d-6f3767cd6346.png)

**Comment by abhigoku10** [comment]
@pandamax  thanks for sharing the idea even i wanted to have a look into it , so as the author suggested did u try the semantic segmentation method for driveable area 
@dbolya  the link you have pointed in your comment can be used for semantic segmentation right ?
i wanted to do lane marking detection also

**Comment by dbolya** [comment]
@abhigoku10 Yeah we already have a semantic segmentation module built in. You just need to expose it like in #73.

**Comment by abhigoku10** [comment]
@pandamax  did u perform the instance segmentation or semantic segmentation for your dirveable are a application , if so what is the code changes made to obtain semantic segmentation

**Comment by abhigoku10** [comment]
@pandamax how many images have you trained you model with to get the output??
@dbolya  to get drive-able and non drive-able area should we train the architecture with label

**Comment by dbolya** [comment]
@abhigoku10 Yeah, train with two instance annotations per image: one for the drivable area, and the other for the non-drivable error. Then, ignore the instance segmentation output of the model and just use the semantic segmentation.

**Comment by pandamax** [comment]
> @pandamax how many images have you trained you model with to get the output??
> @dbolya to get drive-able and non drive-able area should we train the architecture with label
In my experience, I got the datasets from BDD100K, you can use the drivable-area data and convert the data format to cocostyle, here is a [script](https://github.com/ucbdrive/bdd-data/blob/master/bdd_data/bdd2coco.pyl) may help you. And train dataset is 70K, val datasets is 10K.

meanwhle, I opt the instance segmentation method.

**Comment by bond9779** [comment]
hey @pandamax  script which you have  suggested to convert to cocostyle is not available

**Comment by abhigoku10** [comment]
@dbolya  any method on training for these low images , any params we can tune so to see the model generated

**Comment by dbolya** [comment]
@abhigoku10 Wait you're training from scratch from scratch? i.e. not even imagenet pretraining? There are almost as many parameters as pixels in your training data...

You can try changing learning rate and stuff like that, but I doubt you'll get anywhere.

**Comment by abhigoku10** [comment]
@dbolya i am using resent50 imagenet pretrained model for training with kitti 200 images annotated with driveable area as a separate class

**Comment by dbolya** [comment]
@abhigoku10 Right, but you realize that even Mask R-CNN does abysmally on KITTI? There are so few images that the best methods are traditional computer vision hand-crafted algorithms. Machine learning techniques just don't work as well with so few images.

You can try reducing the learning rate, but it's probably just going to overfit.

**Comment by abhigoku10** [comment]
@dbolya @pandamax  i trained the model using BDD100k data set i am getting the exact boxes but the mask is zero , even during training the mask value during validation is zero , but i am getting proper boxes of the road from the model currently the training set consists of 90k images

**Comment by dbolya** [comment]
@abhigoku10 Huh, that shouldn't be happening. Maybe you have a bug somewhere in your config, or if the dataset was not originally in COCO format, a bug in conversion?

**Comment by pandamax** [comment]
@abhigoku10 here is [script](https://github.com/dbolya/yolact/issues/92#issuecomment-531636311) you to check if your data format is cocostyle and visualize your dataset. That may do some help.

**Comment by abhigoku10** [comment]
@pandamax @dbolya  after a through evaluation i had set this (set 'freeze_bn': True in the config). which i made it false now its working on BDD100k dataset , now i shall train on the KITTI dataset and update the results

**Comment by abhigoku10** [comment]
@dbolya @pandamax i have got better results after training using the dataset . I have few queries 
1.the mask value for the road surface detection during validation mAP is low its in the range of 5-10
2. i tried this for lane detection and i get the same the mask value of the range 0-1 
3. Any metrics to validate the output obtained of driveable area and alternate driveable area

**Comment by aungriah** [comment]
@pandamax did you actually manage to add road objects to your direct drivable area segmentor?
